Implement CRUD puppy with interface

patrickmarabeas opened the pull request:
Adds the Puppy, Storer, MapStore and SyncStore type definitions. 

- MapStore implements the Storer interface and supplies a NewMapStore factory function.
- SyncStore implements the Storer interface and supplies a NewSyncStore factory function.
- Uuid increment is handled internally by MapStore Create
- Read will return an empty Puppy struct if provided uuid doesn't exist
- Update will return the given Puppy struct if the uuid doesn't exist
- Destroy returns a bool dependent on uuid existing

Fixes #537

Review of colleague's PR #536

pentaphobe reviewed (commented):
After a while staring at your code (and adding a few extra tests) I cannot work out what’s happening to your coverage.

_However_ one thing which _may_ help is if you conform to the go style of returning an `error` from functions which have an error condition and testing against that.

Still don't think that should matter, but never know...  (also worth mentioning just as a stylistic nitpick)

---

FWIW: I sometimes use the HTML view of coverage to help see what the tool's seeing, but in this case it told me very little.

Either way, this is the command I used:
```bash
go test -v -race -coverprofile /tmp/cover.out -covermode=atomic ./... && go tool cover -html /tmp/cover.out -o /tmp/profile.html && open /tmp/profile.html
```

---

FWIW 2: Tests I tried out were things like:

- doing non deep-equals checks of zero value returns
- calling the offending functions on a `nil` store (which obviously just crashes - alas precluding any coverage result.  Whilst it's arguable that one might need to check for such things, surely that would cause coverage issues on your other functions too)
- started adding a custom error to your code (for my own edification) but the task exploded - and was discarded - once I realised I'd have to update signatures and calls across most files and it's 2am :)

patrickmarabeas commented:
Hopefully it's not error related - as I removed all my error handling upon seeing Lab 07 outline!

anzdaddy commented on 06_puppy/patrickmarabeas/mapStore_test.go:
```diff
@@ -0,0 +1,96 @@
+
+func TestMapStore(t *testing.T) {
+
+	t.Run("NewMapStore", func(t *testing.T) {
```
Seems like these should just be separate top-level `Test…` functions.

patrickmarabeas replied:
Tests have been refactored.

anzdaddy commented on 06_puppy/patrickmarabeas/mapStore_test.go:
```diff
@@ -0,0 +1,96 @@
+	})
+
+	t.Run("Create", func(t *testing.T) {
+		// Setup
```
Avoid comments that state the obvious.

anzdaddy commented on 06_puppy/patrickmarabeas/syncStore.go:
```diff
@@ -0,0 +1,49 @@
+
+// Update modifies the puppy matching the provided uuid in the store with the provided Puppy struct.
+// No action is performed if the identifier does not exist.
+func (store *SyncStore) Update(id int, puppy Puppy) Puppy {
```
Returning the input `Puppy` is redundant, which results in extra testing requirements.

pentaphobe replied:
potential rationale for returning a value: the act of updating potentially has secondary side effects beyond the data in the passed struct (eg. modifying something like `UpdateTime`)

good point re. extra testing req's though

patrickmarabeas replied:
Simplified to bool

anzdaddy commented on 06_puppy/patrickmarabeas/mapStore.go:
```diff
@@ -0,0 +1,50 @@
+
+// Update modifies the puppy matching the provided uuid in the store with the provided Puppy struct.
+// No action is performed if the identifier does not exist.
+func (store *MapStore) Update(id int, puppy Puppy) Puppy {
```
Returning the input `Puppy` is redundant, which results in extra testing requirements.

patrickmarabeas replied:
Simplified to bool.

anzdaddy commented on 06_puppy/patrickmarabeas/syncStore_test.go:
```diff
@@ -0,0 +1,95 @@
+
+func TestSyncStore(t *testing.T) {
+
+	t.Run("NewSyncStore", func(t *testing.T) {
```
Seems like these should just be separate top-level `Test…` functions.

patrickmarabeas replied:
Tests have been refactored.

anzdaddy commented on 06_puppy/patrickmarabeas/syncStore_test.go:
```diff
@@ -0,0 +1,95 @@
+		store := NewMapStore()
+		// Create a puppy
+		id := store.Create(Puppy{Breed: "Wolf", Color: "Grey", Value: "450"})
+		expectedID := 1
```
Is it part of the API design that the first id is 1? If so, then what should the second puppy get after inserting and removing the first Puppy? If not, then you shouldn't test for it.

anzdaddy commented on 06_puppy/patrickmarabeas/syncStore_test.go:
```diff
@@ -0,0 +1,95 @@
+package main
```
Is this a duplicate of `mapStore_test.go` with a different backend? You should implement a single set of parameterised tests.

patrickmarabeas replied:
Tests have been refactored.

juliaogris commented:
Please fix broken build and address @anzdaddy 's comments

patrickmarabeas committed:
Add MapStore which implements the Storer interface

Adds the Puppy, Storer and MapStore type definitions. MapStore implements
the Storer interface and supplies a NewMapStore factory function.

- Uuid increment is handled internally by MapStore Create
- Read will return an empty Puppy struct if provided uuid doesn't exist
- Update returns a bool whether a matching identifier was modified
- Destroy returns a bool whether a matching identifier was modified

Lab 06 (#537)

patrickmarabeas committed:
Add SyncStore which implements the Storer interface

SyncStore implements the Storer interface and supplies a NewSyncStore
factory function.

- Uuid increment is handled internally by SyncStore Create
- Read will return an empty Puppy struct if provided uuid doesn't exist
- Update returns a bool whether a matching identifier was modified
- Destroy returns a bool whether a matching identifier was modified

Lab 06 (#537)

patrickmarabeas committed:
Add empty main function

Add empty main function and tests to pass build requirements

anzboi commented on 06_puppy/patrickmarabeas/types.go:
```diff
@@ -0,0 +1,29 @@
+	Destroy(ID int) bool
+}
+
+type MapStore struct {
```
I would put `MapStore` and `SyncStore` in their respective implementation files `mapStore.go` and `syncStore.go`

patrickmarabeas replied:
Done

anzboi commented on 06_puppy/patrickmarabeas/types.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+type SyncStore struct {
+	uuid int
```
Interesting interpretation of `uuid` you have there.

patrickmarabeas replied:
Changed to `uint`

anzboi commented on 06_puppy/patrickmarabeas/store_test.go:
```diff
@@ -0,0 +1,87 @@
+	a := assert.New(suite.T())
+	id := suite.store.Create(Puppy{Breed: "Boxer", Color: "Brown", Value: "300"})
+
+	a.Equal(id, 1)
```
I think what @anzdaddy was getting at here is that `TestCreate` should create 2 puppies, and check the second ones ID is incremented by 1 from the first, or at least different, think about what could be generating these id's if you don't know what the implementation is.
Otherwise, great tests.

patrickmarabeas replied:
Done

anzboi reviewed (changes requested):
Great work, Just one think I would like to fix up.
From an interface level, we don't care about what ids are being created, only that they are different for any two `Create` calls. IMO the Create test should just create two puppies and check the generated ID's are different.
Feel free to disagree and I'll approve.

rokane commented on 06_puppy/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,5 @@
+package main
+
+func main() {
+
```
I find it a little strange seeing a blank main program. I know the description of the lab doesn't state what you need to include in the main, so you have technically ticked all the boxes, but I feel it makes sense to maybe create a puppy or something to give context to the work being done. Happy to hear your thoughts otherwise.

rokane commented on 06_puppy/patrickmarabeas/mapStore.go:
```diff
@@ -0,0 +1,56 @@
+package main
+
+type MapStore struct {
```
Would be nice to see some documentation on the MapStore struct as well, since it is a part of the exported package.

rokane commented on 06_puppy/patrickmarabeas/syncStore.go:
```diff
@@ -0,0 +1,59 @@
+	"sync"
+)
+
+type SyncStore struct {
```
Again here with the documentation as well

rokane commented on 06_puppy/patrickmarabeas/types.go:
```diff
@@ -0,0 +1,15 @@
+	Value string
+}
+
+type Storer interface {
```
Again here with the documentation.

rokane commented on 06_puppy/patrickmarabeas/store_test.go:
```diff
@@ -0,0 +1,81 @@
+
+func (suite *StoreSuite) TestReadNonExistent() {
+	a := assert.New(suite.T())
+	success := suite.store.Read(100)
```
success seems like a strange name for this variable. Maybe `notFound` or something which indicates that it didn't exist?

rokane reviewed (changes requested):
Looking really good! Great suite of tests and really clean, well documented code. I have left a couple of comments throughout which I think would be nice to have. Also there is the empty main function, if you feel against what I have suggested then let me know, otherwise it would be nice to see something in there (even if it's simple)

juliaogris commented:
The go-course is now closed. Thank you very much for participating.
